File: api/src/main/java/net/blueshell/api/model/ContributionPeriod.java
package net.blueshell.api.model;

import jakarta.persistence.*;
import lombok.Data;
import lombok.Setter;
import net.blueshell.api.base.BaseModel;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.sql.Timestamp;
import java.time.LocalDate;
import java.util.Objects;
import java.util.Set;

@Data
@Entity
@Table(name = "contribution_periods")
@SQLDelete(sql = "UPDATE contribution_periods SET deleted_at = CURRENT_TIMESTAMP WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
public class ContributionPeriod  implements BaseModel<Long> {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToMany(mappedBy = "contributionPeriod", cascade = CascadeType.ALL)
    private Set<Contribution> contributions;

    @Column(name = "start_date")
    @Setter
    private LocalDate startDate;

    @Column(name = "end_date")
    @Setter
    private LocalDate endDate;

    @Column(name = "half_year_fee")
    @Setter
    private double halfYearFee;

    @Column(name = "full_year_fee")
    @Setter
    private double fullYearFee;

    @Column(name = "alumni_fee")
    @Setter
    private double alumniFee;

    @Column(name = "list_id")
    @Setter
    private Long listId;

    @Column(name = "deleted_at")
    private Timestamp deletedAt;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ContributionPeriod that = (ContributionPeriod) o;
        return id != null && Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: api/src/main/kotlin/net/blueshell/api/contribution/domain/model/ContributionPeriod.kt
package net.blueshell.api.contribution.domain.model

import jakarta.persistence.*
import net.blueshell.api.shared.jpa.JpaListener
import net.blueshell.api.shared.model.AuditedAutoIdEntity
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction
import java.time.LocalDate

@Entity
@Table(
    name = "contribution_periods",
    uniqueConstraints = [
        UniqueConstraint(
            name = "uk_contribution_periods_start_end_deleted_at",
            columnNames = ["start_date", "end_date", "deleted_at"]
        )
    ],
    indexes = [
        Index(name = "idx_contribution_periods_deleted_at", columnList = "deleted_at"),
        Index(name = "idx_contribution_periods_start_date", columnList = "start_date"),
        Index(name = "idx_contribution_periods_end_date", columnList = "end_date"),
        Index(name = "idx_contribution_periods_list_id", columnList = "list_id")
    ]
)
@SQLDelete(sql = "UPDATE contribution_periods SET deleted_at = NOW(), version = version + 1 WHERE id = ? AND version = ?")
@SQLRestriction("deleted_at = '9999-12-31 23:59:59'")
@EntityListeners(JpaListener::class)
class ContributionPeriod : AuditedAutoIdEntity() {
    @OneToMany(mappedBy = "_contributionPeriod", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    private val _contributions: MutableSet<Contribution> = linkedSetOf()
    val contributions: Set<Contribution>
        get() = _contributions

    @Column(name = "start_date", nullable = false)
    lateinit var startDate: LocalDate

    @Column(name = "end_date")
    lateinit var endDate: LocalDate

    @Column(name = "half_year_fee", nullable = false)
    var halfYearFee = 0.0

    @Column(name = "full_year_fee", nullable = false)
    var fullYearFee = 0.0

    @Column(name = "alumni_fee", nullable = false)
    var alumniFee = 0.0

    @Column(name = "list_id")
    var listId: Long? = null
}
